### PR TITLE
Add AbandonState exception

### PR DESCRIPTION
**Pull request by ex0dus-0x**
Fixes https://github.com/trailofbits/manticore/issues/1103.

_(BTW, this is Alan, the new intern)_

---
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/1222)

### CODE REVIEW

File: manticore/core/state.py [hunk 1]
```diff
@@ -196,7 +201,7 @@ def abandon(self):
 
         Note: This must be called from the Executor loop, or a :func:`~manticore.Manticore.hook`.
         '''
-        raise TerminateState("Abandoned state")
+        raise AbandonState
```
Comment (stephan-tolksdorf) [inline manticore/core/state.py]: `AbandonState` inherits the `__init__` method from `TerminateState`, so this line misses the required string argument and hence will throw a different exception. One way to fix this issue would be to give `AbandonState` its own `__init__` method with a message argument that has the default value `'Abandoned state'`.

**Comment by stephan-tolksdorf** [comment]
@disconnect3d @ex0dus-0x The commit actually causes [an error](https://travis-ci.org/trailofbits/manticore/jobs/445842911) in an example script, but since the `Executor` (logs and) suppresses all exceptions, this isn't recognized as an error by the test runner.

**Comment by disconnect3d** [comment]
@stephan-tolksdorf Thanks for spotting this. I've reverted it in https://github.com/trailofbits/manticore/pull/1225

@ex0dus-0x can you fix it and open a new PR?

File: manticore/core/state.py (hunk 1 above)
Comment (disconnect3d) [reply manticore/core/state.py]: :+1:

```python
In [1]: class A(Exception): pass

In [2]: class B(A):
   ...:     def __init__(self, msg):
   ...:         pass
   ...:

In [3]: class C(B): pass

In [4]:

In [4]: raise C
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-4a531f008761> in <module>()
----> 1 raise C

TypeError: __init__() missing 1 required positional argument: 'msg'
```

**Pull request merged**